### PR TITLE
Remove __pthread_detached_exit in favor of __emscripten_thread_cleanup. NFC.

### PR DESCRIPTION
**Pull request by kleisauke**
Split from PR #13007.

### CODE REVIEW

File: src/library_pthread.js [hunk 1]
```diff
@@ -309,7 +304,7 @@ var LibraryPThread = {
         worker.on('error', function(e) {
           worker.onerror(e);
         });
-        worker.on('detachedExit', function() {
+        worker.on('exit', function() {
```
Comment (sbc100) [inline src/library_pthread.js:307]: Why this change rather than just removing this handler completely like above?  What is the 'exit' message here?

File: system/lib/pthread/pthread_create.c [hunk 2]
```diff
@@ -236,7 +236,7 @@ void _emscripten_thread_exit(void* result) {
   // object and we are done.
   if (state == DT_DETACHED) {
     self->detach_state = DT_EXITED;
-    __pthread_detached_exit();
+    __emscripten_thread_cleanup(self);
```
Comment (sbc100) [inline system/lib/pthread/pthread_create.c:239]: I was thinking about doing this .. but I was slightly worried about free'ing the stack being used by the current thread.   

Seems a little dangerous.  I don't come up with better solution yet but I was thinking maybe we could instead call this from worker.js?  (i.e.  once the stack is unwound all the way)?

File: src/library_pthread.js (hunk 1 above)
Comment (kleisauke) [reply src/library_pthread.js:307]: The `exit` -> `detachedExit` change was mistakenly introduced in PR #12861 (see comment https://github.com/emscripten-core/emscripten/pull/12861#discussion_r700102217), but perhaps it is indeed better to remove this listener completely (since issue #9763 may have already been resolved, [`test_pthread_create.cpp`](https://github.com/emscripten-core/emscripten/blob/main/tests/pthread/test_pthread_create.cpp) runs fine on Node).

Comment (sbc100) [reply src/library_pthread.js:307]: Oh I see.  Can you split that out into its own fix and mention that is address an accidental change from #12861?  Thanks for doing all this work BTW, its much appreciated cleanup.

File: system/lib/pthread/pthread_create.c (hunk 2 above)
Comment (kleisauke) [reply system/lib/pthread/pthread_create.c:239]: IIUC, `__emscripten_thread_free_data` was also called by the `detachedExit`/`__pthread_detached_exit`, so this PR should be completely non-functional (except that it now sends the `cleanupThread` command, which will also eventually call `returnWorkerToPool`/`__emscripten_thread_free_data`).

File: src/library_pthread.js (hunk 1 above)
Comment (kleisauke) [reply src/library_pthread.js]: No problem! I've reverted this change with commit d69ff5bb431ce13a815b12989bb78fad5ae1dd0f, and will make a new PR with that (or one that removes this listener completely).

File: system/lib/pthread/pthread_create.c (hunk 2 above)
Comment (sbc100) [reply system/lib/pthread/pthread_create.c:239]: IIUC, In the current code, when `__pthread_detached_exit` is called the `cleanupThread` command is sent back to the main thread which ends up freeing the data, so the data is not freed on the exit'ing thread at all.

Comment (sbc100) [reply system/lib/pthread/pthread_create.c:239]: Also, with this new PR, how is the detached thread ever returned to the worker pool?

Comment (sbc100) [reply system/lib/pthread/pthread_create.c:239]: Oh wait, I see that `__emscripten_thread_cleanup` also posts a message back to the main thread.. so I agree this should be and NFC.  Thanks for clarifying!

**Review by sbc100 — APPROVED** [review]

**Comment by kleisauke** [comment]
Is there anything needed to land this? I could do this myself, but I'm a little uncomfortable with merging PRs as an external contributor.
